File: 39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python
#
# [39] 组合总和
#

# @lc code=start
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def foo(candis, target, tmp, res):
            if target == 0:
                if sorted(tmp) not in res:
                    res.append(sorted(tmp))
                return
            
            if target < 0:
                return

            for i in candis:
                nxt = tmp[:]
                nxt += [i]
                foo(candis, target - i, nxt, res)
        
        res = []
        foo(candidates, target, [], res)
        return res
# @lc code=end

File: 39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
public:
    set<vector<int>> res;
    void foo(vector<int>& candidates, int target, const vector<int>& current) {
        if (target == 0) {
            if (!current.empty()) {
                auto tmp = current;
                std::sort(tmp.begin(), tmp.end());
                res.insert(tmp);
            }
            return;
        }
        if (target < 0) {
            return;
        }
        for (const auto& i : candidates) {
            auto next = current;
            next.push_back(i);
            foo(candidates, target - i, next);
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        foo(candidates, target, {});
        vector<vector<int>> r;
        for (auto& i : res) {
            r.push_back(i);
        }
        return r;
    }
};
// @lc code=end

